File: app/src/main/java/com/kotobyte/searchpage/SearchPagePresenter.java
package com.kotobyte.searchpage;

import com.kotobyte.models.Kanji;
import com.kotobyte.models.WordLiteral;
import com.kotobyte.models.Word;

import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

class SearchPagePresenter implements SearchPageContracts.Presenter {

    private SearchPageContracts.View mView;
    private SearchPageContracts.DataSource mDataSource;
    private String mSearchQuery;

    private Scheduler mBackgroundScheduler = Schedulers.io();
    private Scheduler mMainThreadScheduler = AndroidSchedulers.mainThread();
    private Disposable mWordSearchOperationSubscription;
    private Disposable mKanjiSearchOperationSubscription;

    SearchPagePresenter(
            SearchPageContracts.View view,
            SearchPageContracts.DataSource dataSource,
            String searchQuery) {

        mView = view;
        mDataSource = dataSource;
        mSearchQuery = searchQuery;
    }

    @Override
    public void onCreate() {
        searchWords();
    }

    @Override
    public void onDestroy() {

        if (mWordSearchOperationSubscription != null) {
            mWordSearchOperationSubscription.dispose();
        }

        if (mKanjiSearchOperationSubscription != null) {
            mKanjiSearchOperationSubscription.dispose();
        }
    }

    @Override
    public void onRequestKanjiListForWord(int position, Word word) {
        searchKanjiForWord(position, word);
    }

    @Override
    public void onRequestDetailForKanji(Kanji kanji) {
        mView.showKanjiDetailScreen(kanji);
    }

    private void searchWords() {

        mView.showWordSearchProgressBar(true);
        mView.showNoWordSearchResultsLabel(false);
        mView.showWordSearchResultsView(false);

        mWordSearchOperationSubscription = mDataSource.searchWords(mSearchQuery)
                .subscribeOn(mBackgroundScheduler)
                .observeOn(mMainThreadScheduler)
                .subscribe(new Consumer<List<Word>>() {

                    @Override
                    public void accept(@NonNull List<Word> words) throws Exception {

                        mView.showWordSearchProgressBar(false);

                        if (words.isEmpty()) {
                            mView.showNoWordSearchResultsLabel(true);

                        } else {
                            mView.showWordSearchResults(words);
                            mView.showWordSearchResultsView(true);
                        }

                    }

                }, new Consumer<Throwable>() {

                    @Override
                    public void accept(@NonNull Throwable throwable) throws Exception {

                        mView.showWordSearchProgressBar(false);
                        mView.showUnknownError(throwable);
                    }
                });
    }

    private void searchKanjiForWord(final int position, final Word word) {

        if (mKanjiSearchOperationSubscription != null) {
            mKanjiSearchOperationSubscription.dispose();
        }

        mKanjiSearchOperationSubscription = mDataSource.searchKanji(getKanjiSearchQueryForWord(word))
                .subscribeOn(mBackgroundScheduler)
                .observeOn(mMainThreadScheduler)
                .delaySubscription(100, TimeUnit.MILLISECONDS)
                .subscribe(new Consumer<List<Kanji>>() {

                    @Override
                    public void accept(@NonNull List<Kanji> kanjiList) throws Exception {

                        mView.showKanjiSearchResults(position, kanjiList);
                    }

                }, new Consumer<Throwable>() {

                    @Override
                    public void accept(@NonNull Throwable throwable) throws Exception {

                        mView.showUnknownError(throwable);
                    }
                });
    }

    private String getKanjiSearchQueryForWord(Word word) {
        StringBuilder queryBuilder = new StringBuilder();

        for (WordLiteral wordLiteral : word.getLiterals()) {
            queryBuilder.append(wordLiteral.getText());
        }

        return queryBuilder.toString();
    }
}

File: app/src/main/kotlin/com/kotobyte/search/SearchPagePresenter.kt
package com.kotobyte.search

import com.kotobyte.models.Kanji
import com.kotobyte.models.Word
import java.util.concurrent.TimeUnit

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

internal class SearchPagePresenter(
        private val view: SearchPageContracts.View,
        private val dataSource: SearchPageContracts.DataSource,
        private val searchQuery: String) : SearchPageContracts.Presenter {

    private val backgroundScheduler = Schedulers.io()
    private val mainThreadScheduler = AndroidSchedulers.mainThread()

    private var wordSearchSubscription: Disposable? = null
    private var kanjiSearchSubscription: Disposable? = null

    override fun onCreate() = searchWords()

    override fun onDestroy() {

        wordSearchSubscription?.dispose()
        kanjiSearchSubscription?.dispose()
    }

    override fun onRequestKanjiListForWord(position: Int, word: Word) =
            searchKanjiForWord(position, word)

    override fun onRequestDetailForKanji(kanji: Kanji) = view.showKanjiDetailScreen(kanji)

    private fun searchWords() {

        view.showWordSearchProgressBar(true)
        view.showNoWordSearchResultsLabel(false)
        view.showWordSearchResultsView(false)

        wordSearchSubscription = dataSource.searchWords(searchQuery)
                .subscribeOn(backgroundScheduler)
                .observeOn(mainThreadScheduler)

                .subscribe({ words ->
                    view.showWordSearchProgressBar(false)

                    if (words.isEmpty()) {
                        view.showNoWordSearchResultsLabel(true)

                    } else {
                        view.showWordSearchResults(words)
                        view.showWordSearchResultsView(true)
                    }

                }) { throwable ->
                    view.showWordSearchProgressBar(false)
                    view.showUnknownError(throwable)
                }
    }

    private fun searchKanjiForWord(position: Int, word: Word) {

        kanjiSearchSubscription?.dispose()

        kanjiSearchSubscription = dataSource.searchKanji(getKanjiSearchQueryForWord(word))
                .subscribeOn(backgroundScheduler)
                .observeOn(mainThreadScheduler)
                .delaySubscription<Any>(200, TimeUnit.MILLISECONDS)

                .subscribe({ kanjiList ->
                    view.showKanjiSearchResults(position, kanjiList)

                }) { throwable ->
                    view.showUnknownError(throwable)
                }
    }

    private fun getKanjiSearchQueryForWord(word: Word): String {
        val queryBuilder = StringBuilder()

        for ((text) in word.literals) {
            queryBuilder.append(text)
        }

        return queryBuilder.toString()
    }
}
